cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -161,23 +163,60 @@ struct CastFunctor<T, BooleanType, enable_if_number<T>> {
   }
 };
 
+// Number to Boolean
+template <typename I>
+struct CastFunctor<BooleanType, I,
+                   typename std::enable_if<std::is_base_of<Number, I>::value &&
+                                           !std::is_same<BooleanType, I>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    auto in_data = input.GetValues<typename I::c_type>(1);
+    const auto generate = [&in_data]() -> bool { return *in_data++ != 0; };
+    internal::GenerateBitsUnrolled(output->buffers[1]->mutable_data(), output->offset,
+                                   input.length, generate);
+  }
+};
+
 // ----------------------------------------------------------------------
 // Integers and Floating Point
 
+// Conversions pairs (<O, I>) are partitioned in 4 type traits:
+// - is_number_downcast
+// - is_integral_signed_to_unsigned
+// - is_integral_unsigned_to_signed
+// - is_float_truncate
+//
+// Each class has a different way of validation if the conversion is safe
+// (either with bounded intervals or with explicit C casts)
+
+template <typename O, typename I, typename Enable = void>
+struct is_number_downcast {
+  static constexpr bool value = false;
+};
+
 template <typename O, typename I>
-struct is_numeric_cast {
+struct is_number_downcast<
+    O, I,
+    typename std::enable_if<std::is_base_of<Number, O>::value &&
+                            std::is_base_of<Number, I>::value>::type> {
+  using O_T = typename O::c_type;
+  using I_T = typename I::c_type;
+
   static constexpr bool value =
-      (std::is_base_of<Number, O>::value && std::is_base_of<Number, I>::value) &&
-      (!std::is_same<O, I>::value);
+      ((!std::is_same<O, I>::value) &&
+       ((std::is_signed<O_T>::value == std::is_signed<I_T>::value) &&
+        (std::is_floating_point<O_T>::value == std::is_floating_point<I_T>::value)) &&
```

Comment:
Seems this constexpr clause is only true when both input and output are floating point. Should the previous && be ||?

Reply:
It's true when both Input and Output are of the same floating point (i.e. both floating or both integrals), note the use of `==` instead of `&&`.

Reply:
Got it, reading again with fresh eyes looks fine =)